### PR TITLE
Updates to make CGATPipelines Python 3 compatable

### PR DESCRIPTION
**Pull request by TomSmithCGAT**
Most of the changes were to resolve errors I got when running the pipelines with python 3.5. Some of these definitely need looking over by somebody else before they get merged into master:

-  CGATPipelines/Pipeline/Parameters.py: @AndreasHeger I was getting errors relating to '%' in my pipeline.ini parameter values. To my knowledge we never use the '%' for string subsitution within pipeline.ini so I assume we're better off using the RawConfigParser? The other option is to escape the '%', i.e '%%'. I never had this issue with py2 so I assume it's a py3-specific problem?

- Lots of instances where a string needs to be encoded to bytes for writing out etc. I've tested small chunks of this code out in py2 to check I get the same result but it would be good to run the regression testing too. What's the best way to run the CGATPipelines regression testing in py2.7  & py3.5?

### CODE REVIEW

File: CGATPipelines/PipelineMapping.py [hunk 1]
```diff
@@ -313,7 +313,7 @@ def mergeAndFilterGTF(infile, outfile, logfile,
         new_exons.sort(key=lambda x: (x.start, x.gene_id, x.transcript_id))
 
         for e in new_exons:
-            outf.write("%s\n" % str(e))
+            outf.write(("%s\n" % str(e)).encode())
```
Comment (sebastian-luna-valero) [inline CGATPipelines/PipelineMapping.py]: Thanks @TomSmithCGAT 

Do we need `.encode('utf-8')` here as well?

**Comment by sebastian-luna-valero** [comment]
The regression tests for pipelines are currently running locally with Jenkins, due to the computational resources they need. We are testing so far: ancestral_repeats, annotations, intervals, mapping, peakcalling, readqc, rnaseqdiffexpression, rnaseqqc, windows. These tests are Py2 only for the time being but I will be looking at Py3 testing soon.

The plan is to include to the rest of the "production" pipelines to Jenkins once we short-list them.

Testing in Jenkins has to be improved as well. Not only to support Py3 but to run tests on github branches (which is not the case right now). Currently, it runs regression tests on the master branch overnight. I can also give you access to Jenkins so you can check the results of the tests and you will be able to trigger these regression tests on demand.

**Comment by sebastian-luna-valero** [comment]
On second thoughts, you can run regression tests on demand by doing:
```
# make sure you are on your branch
cd <CGATPipelines-clone>
git checkout <your-branch>

# run tests
mkdir <new-folder-for-testing>; cd <new-folder-for-testing>
wget https://raw.githubusercontent.com/CGATOxford/CGATTests/master/pipeline.ini
wget https://raw.githubusercontent.com/CGATOxford/CGATTests/master/conf.py
python <CGATPipelines-clone>/CGATPipelines/pipeline_testing.py make full
python <CGATPipelines-clone>/CGATPipelines/pipeline_testing.py make build_report
```

By the way, this is what Jenkins does locally overnight on the `master` branch. 

The commands above will run all regression tests for these pipelines: ancestral_repeats, annotations, intervals, mapping, peakcalling, readqc, rnaseqdiffexpression, rnaseqqc, windows, and enrichment. If you only want to run regression tests for a specific pipeline, then edit the pipeline.ini and leave only the section for the pipeline that you want to test.

You can also query the generated `csvdb` in order to see the results:
```
sqlite3 <new-folder-for-testing>/csvdb "SELECT status FROM md5_compare"
```

I hope that helps.

**Comment by TomSmithCGAT** [comment]
Great! Thanks Sebastian.

**Comment by AndreasHeger** [comment]
@TomSmithCGAT , not sure about RawConfigParser, the interpolation feature in ConfigParser might be useful? Maybe SafeConfigParser is better.

**Comment by sebastian-luna-valero** [comment]
Hi Tom,

I have found problems with `pipeline_ancestral_repeats.py` when running the regression tests on this branch. Please see the relevant output below.

The problem is caused in the `exportRatesAsGFF` task:
https://github.com/CGATOxford/CGATPipelines/blob/%7BTS%7D-Py35/CGATPipelines/pipeline_ancestral_repeats.py#L554

when loading `PARAMS["interface_rates_query_gff"]` from its `pipeline.ini`:
https://github.com/CGATOxford/CGATPipelines/blob/%7BTS%7D-Py35/CGATPipelines/pipeline_ancestral_repeats/pipeline.ini#L74

According to a quick `grep` over this repository shows that  `pipeline_ancestral_repeats.py` is the only pipeline using `%` for string substitution within `pipeline.ini`. @AndreasHeger how do you suggest to proceed?

Best regards,
Sebastian

```
2017-02-09 08:22:45,716 DEBUG Experiment.debug.1015 running statement:
                                                    gunzip     < aligned_repeats.rates.gz     | cgat csv_cut qName qStart qEnd distance converged     | awk '!/qName/ && $5 {printf("%s\tancestral_repeat\texon\t%s\t%s\t%s\t+\t.\t.\n", $1, $2, $3, $4);}'     | gzip     > %(query)s_rates.gff.gz
2017-02-09 08:22:45,717 DEBUG Experiment.debug.1015 Job spec is: -V -N _hg19_rates.gff.gz -p -10 -l mem_free=2G -q all.q
2017-02-09 08:22:45,736 DEBUG Experiment.debug.1015 job has been submitted with job_id 5721435
2017-02-09 08:22:56,761 ERROR Experiment.error.1020 1 tasks with errors, please see summary below:
2017-02-09 08:22:56,762 ERROR Experiment.error.1020 0: Task=exportRatesAsGFF Error=exceptions.OSError Job=[aligned_repeats.rates.gz->%(query)s_rates.gff.gz]: 
2017-02-09 08:22:56,762 ERROR Experiment.error.1020 full traceback is in pipeline.log
2017-02-09 08:22:56,762 ERROR Control.main.1011 start of error messages
2017-02-09 08:22:56,762 ERROR Control.main.1012 
                                                    
                                                    
                                                    
                                                Original exception:
                                                
                                                    Exception #1
                                                      'exceptions.OSError(---------------------------------------
                                                    Child was terminated by signal 2: 
                                                    The stderr was: 
                                                    /ifs/projects/sebastian/2017-02-03-pipeline-testing-workshop/test_ancestral_repeats.dir/ctmpikpEpv: line 29: syntax error near unexpected token `('
                                                    
                                                    gunzip     < aligned_repeats.rates.gz     | cgat csv_cut qName qStart qEnd distance converged     | awk '!/qName/ && $5 {printf("%s\tancestral_repeat\texon\t%s\t%s\t%s\t+\t.\t.\n", $1, $2, $3, $4);}'     | gzip     > %(query)s_rates.gff.gz
                                                    -----------------------------------------)' raised in ...
                                                       Task = def exportRatesAsGFF(...):
                                                       Job  = [aligned_repeats.rates.gz -> %(query)s_rates.gff.gz]
                                                    
                                                    Traceback (most recent call last):
                                                      File "/ifs/apps/apps/python-2.7.11-gcc-5.2.1/lib/python2.7/site-packages/ruffus/task.py", line 751, in run_pooled_job_without_exceptions
                                                        register_cleanup, touch_files_only)
                                                      File "/ifs/apps/apps/python-2.7.11-gcc-5.2.1/lib/python2.7/site-packages/ruffus/task.py", line 567, in job_wrapper_io_files
                                                        ret_val = user_defined_work_func(*params)
                                                      File "/ifs/devel/sebastian/CGATPipelines/CGATPipelines/pipeline_ancestral_repeats.py", line 565, in exportRatesAsGFF
                                                        P.run()
                                                      File "/ifs/devel/sebastian/CGATPipelines/CGATPipelines/Pipeline/Execution.py", line 570, in run
                                                        ignore_errors=ignore_errors)
                                                      File "/ifs/devel/sebastian/CGATPipelines/CGATPipelines/Pipeline/Cluster.py", line 348, in collectSingleJobFromCluster
                                                        "".join(stderr), statement))
                                                    OSError: ---------------------------------------
                                                    Child was terminated by signal 2: 
                                                    The stderr was: 
                                                    /ifs/projects/sebastian/2017-02-03-pipeline-testing-workshop/test_ancestral_repeats.dir/ctmpikpEpv: line 29: syntax error near unexpected token `('
                                                    
                                                    gunzip     < aligned_repeats.rates.gz     | cgat csv_cut qName qStart qEnd distance converged     | awk '!/qName/ && $5 {printf("%s\tancestral_repeat\texon\t%s\t%s\t%s\t+\t.\t.\n", $1, $2, $3, $4);}'     | gzip     > %(query)s_rates.gff.gz
                                                    -----------------------------------------
                                                    
                                                    
2017-02-09 08:22:56,762 ERROR Control.main.1013 end of error messages
```

**Comment by AndreasHeger** [comment]
Will need to look at this in detail. We need to make good use of six.py and other utilities to make changes. 

For the %, this is a useful functionality. Don't know what exactly the error is, but maybe use SafeConfigParser?

**Comment by AndreasHeger** [comment]
Hi @TomSmithCGAT  and @sebastian-luna-valero , I think my take on this pull-request is to hold off with making the py3 compatibility changes now. Let us wait until we have all in one place. For example, I am bit hesitant to add all these .encode() calls everywhere, as these likely need to have a fundamental solution. For example, the getTempFile() could except an encoding option. Also, encoding needs to be set-able globally, hardcoding UTF-8 will cause issue later.

@TomSmithCGAT, is it ok to remove all the py3 fixes, but keep any other fixes you have made in this PR?

**Comment by TomSmithCGAT** [comment]
Hi @AndreasHeger. That's fine by me.

**Comment by sebastian-luna-valero** [comment]
Thanks @AndreasHeger and @TomSmithCGAT 

I have reverted the Py3 changes. If you are happy with the changes, I will merge with master.

EDIT: All tests passing locally on Py2

**Comment by AndreasHeger** [comment]
Thanks, looks good!

**Pull request merged**